file/file.go

```diff
@@ -0,0 +1,166 @@
+package file
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/ipld/go-ipld-prime"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+)
+
+// NewUnixFSFile attempts to construct an ipld node from the base protobuf node representing the
+// root of a unixfs File.
+// It provides a `bytes` view over the file, along with access to io.Reader streaming access
+// to file data.
+func NewUnixFSFile(ctx context.Context, substrate ipld.Node, lsys *ipld.LinkSystem) (StreamableByteNode, error) {
+	if substrate.Kind() == ipld.Kind_Bytes {
+		// A raw / single-node file.
+		return &singleNodeFile{substrate, 0}, nil
+	}
+	return &shardNodeFile{ctx, lsys, substrate, nil}, nil
+}
+
+// A StreamableByteNode is an ipld.Node that can be streamed over. It is guaranteed to have a Bytes type.
+type StreamableByteNode interface {
+	ipld.Node
+	io.Reader
+}
+
+type singleNodeFile struct {
+	ipld.Node
+	ptr int
+}
+
+func (f *singleNodeFile) Read(p []byte) (int, error) {
+	buf, err := f.Node.AsBytes()
+	if err != nil {
+		return 0, err
+	}
+	if f.ptr >= len(buf) {
+		return 0, io.EOF
+	}
+	n := copy(p, buf[f.ptr:])
+	f.ptr += n
+	return n, nil
+}
+
+type shardNodeFile struct {
+	ctx       context.Context
+	lsys      *ipld.LinkSystem
+	substrate ipld.Node
+	rdr       io.Reader
+}
+
+var _ ipld.Node = (*shardNodeFile)(nil)
+
+func (s *shardNodeFile) Read(p []byte) (int, error) {
+	// collect the sub-nodes on first use
+	if s.rdr == nil {
+		links, err := s.substrate.LookupByString("Links")
+		if err != nil {
+			return 0, err
+		}
+		readers := make([]io.Reader, 0)
+		lnki := links.ListIterator()
+		for !lnki.Done() {
+			_, lnk, err := lnki.Next()
+			if err != nil {
+				return 0, err
+			}
+			if pbl, ok := lnk.(dagpb.PBLink); ok {
+				target, err := s.lsys.Load(ipld.LinkContext{Ctx: s.ctx}, pbl.Hash.Link(), basicnode.Prototype.Any)
```

Comment:
This could probably use a prototype that's specific to dag-pb structures, since we know it's that all the way down in unixfsv1?  (Or possibly a separate prototype for bytes, sometimes, if the CID indicates raw mode?  I'm not familiar with exactly how that part of the protocol works; maybe the dag-pb structures already account for that.)

Reply:
Leaf nodes can potentially be bytes, depending on the type of file encoding. We should know off of the cid codec.
I wonder if there's a way to avoid having this type of logic every time the link system is used?